crates/pathfinder/src/rpc/v02/method/get_transaction_by_block_id_and_index.rs

```diff
@@ -0,0 +1,256 @@
+use anyhow::Context;
+
+use crate::core::{BlockId, StarknetTransactionIndex};
+use crate::rpc::v02::types::reply::Transaction;
+use crate::rpc::v02::RpcContext;
+use crate::storage::{StarknetBlocksBlockId, StarknetBlocksTable, StarknetTransactionsTable};
+
+#[derive(serde::Deserialize, Debug, PartialEq, Eq)]
+pub struct GetTransactionByBlockIdAndIndexInput {
+    block_id: BlockId,
+    index: StarknetTransactionIndex,
+}
+
+crate::rpc::error::generate_rpc_error_subset!(
+    GetTransactionByBlockIdAndIndexError: BlockNotFound,
+    InvalidTxnIndex
+);
+
+#[allow(dead_code)]
+pub async fn get_transaction_by_block_id_and_index(
+    context: std::sync::Arc<RpcContext>,
+    input: GetTransactionByBlockIdAndIndexInput,
+) -> Result<Transaction, GetTransactionByBlockIdAndIndexError> {
+    let index: usize = input
+        .index
+        .get()
+        .try_into()
+        .map_err(|_| GetTransactionByBlockIdAndIndexError::InvalidTxnIndex)?;
+
+    let block_id = match input.block_id {
+        BlockId::Hash(hash) => hash.into(),
+        BlockId::Number(number) => number.into(),
+        BlockId::Latest => StarknetBlocksBlockId::Latest,
+        BlockId::Pending => {
+            return get_transaction_from_pending(&context.pending_data, index).await
+        }
+    };
+
+    let storage = context.storage.clone();
+    let span = tracing::Span::current();
+
+    let jh = tokio::task::spawn_blocking(move || {
+        let _g = span.enter();
+        let mut db = storage
+            .connection()
+            .context("Opening database connection")?;
+
+        let db_tx = db.transaction().context("Creating database transaction")?;
+
+        // Get the transaction from storage.
+        match StarknetTransactionsTable::get_transaction_at_block(&db_tx, block_id, index)
+            .context("Reading transaction from database")?
+        {
+            Some(transaction) => Ok(transaction.into()),
+            None => {
+                // We now need to check whether it was the block hash or transaction index which were invalid. We do this by checking if the block exists
+                // at all. If no, then the block hash is invalid. If yes, then the index is invalid.
+                //
+                // get_root is cheaper than querying the full block.
+                match StarknetBlocksTable::get_root(&db_tx, block_id)
+                    .context("Reading block from database")?
+                {
+                    Some(_) => Err(GetTransactionByBlockIdAndIndexError::InvalidTxnIndex),
+                    None => Err(GetTransactionByBlockIdAndIndexError::BlockNotFound),
+                }
```

Comment:
Opportunity for some custom sql instead. Or we could punt that exercise for once we've remove `v0.1` and do an audit over all of these storage methods.

Reply:
Yeah, was considering that -- finding out why the query failed with another query is awkward. It't be great to factor out the actual decompressing & converting of the transaction JSON to a proper type and then create a custom query here?
I'll check how feasible this is.

Reply:
Unfortunately that query would probably just make things worse. We now have a query matching on the (block hash, index) tuple on the transactions table, plus another one looking up the block by hash -- both extremely simple (and fast given the indices we have on these tables). On the other hand, a custom query doing both would basically be just a union of the two queries.

Reply:
Okay, no problem. Multiple simple queries are also acceptable.